File: packages/android/core-domain/src/main/kotlin/app/notes/domain/Note.kt
package app.notes.domain

import java.time.Instant

data class Note(
    val id: NoteId,
    val ownerId: UserId,
    val notebookId: NotebookId?,
    val title: String,
    val bodyJson: String,
    val tags: Set<TagId>,
    val createdAt: Instant,
    val updatedAt: Instant,
    val isPinned: Boolean,
    val isTrashed: Boolean,
    val trashedAt: Instant?,
    val aiExcluded: Boolean,
    val attachmentRefs: List<AttachmentRef>,
) {
    init {
        require(title.isNotBlank()) { "Note.title cannot be empty" }
        require(bodyJson.isJsonParseable()) { "Note.bodyJson must be valid JSON" }
        if (isTrashed) {
            require(trashedAt != null) { "Note.trashedAt must be set when isTrashed is true" }
        } else {
            require(trashedAt == null) { "Note.trashedAt must be null when isTrashed is false" }
        }
    }
}

/**
 * Lightweight JSON validity probe — we don't pull in a JSON dependency in
 * core-domain, so the check is structural: balanced braces / brackets,
 * starts with `{`/`[`/string/number/literal. The full parser lives in the
 * editor module that actually consumes the body.
 */
internal fun String.isJsonParseable(): Boolean {
    val trimmed = trim()
    if (trimmed.isEmpty()) return false
    val first = trimmed.first()
    if (first !in setOf('{', '[', '"', 't', 'f', 'n') && !first.isDigit() && first != '-') {
        return false
    }
    var depthCurly = 0
    var depthSquare = 0
    var inString = false
    var escape = false
    for (c in trimmed) {
        if (inString) {
            when {
                escape -> escape = false
                c == '\\' -> escape = true
                c == '"' -> inString = false
            }
            continue
        }
        when (c) {
            '"' -> inString = true
            '{' -> depthCurly++
            '}' -> if (--depthCurly < 0) return false
            '[' -> depthSquare++
            ']' -> if (--depthSquare < 0) return false
        }
    }
    return !inString && depthCurly == 0 && depthSquare == 0
}

File: packages/domain/src/note.ts
import type { NoteId, NotebookId, TagId, UserId } from "./ids.ts";
import type { AttachmentRef } from "./attachment.ts";
import { assertIsoTimestamp } from "./notebook.ts";

export interface Note {
  readonly id: NoteId;
  readonly ownerId: UserId;
  readonly notebookId: NotebookId | null;
  readonly title: string;
  readonly bodyJson: string;
  readonly tags: ReadonlySet<TagId>;
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly isPinned: boolean;
  readonly isTrashed: boolean;
  readonly trashedAt: string | null;
  readonly aiExcluded: boolean;
  readonly attachmentRefs: readonly AttachmentRef[];
}

export function makeNote(input: Note): Note {
  if (input.title.trim().length === 0) {
    throw new Error("Note.title cannot be empty");
  }
  try {
    JSON.parse(input.bodyJson);
  } catch {
    throw new Error("Note.bodyJson must be valid JSON");
  }
  assertIsoTimestamp("Note.createdAt", input.createdAt);
  assertIsoTimestamp("Note.updatedAt", input.updatedAt);
  if (input.isTrashed) {
    if (input.trashedAt === null) {
      throw new Error("Note.trashedAt must be set when isTrashed is true");
    }
    assertIsoTimestamp("Note.trashedAt", input.trashedAt);
  } else if (input.trashedAt !== null) {
    throw new Error("Note.trashedAt must be null when isTrashed is false");
  }
  return input;
}
